### ISSUE TITLE
FunctionList fails if removing before adding

### ISSUE THREAD

**Issue by gorenje** [supercollider/supercollider#5935]
## Environment

* SuperCollider version: 3.12.2
* Operating system: MacOS 10.13
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce

```supercollider
*** Welcome to SuperCollider 3.12.2. *** For help type cmd-d.
2022-12-16 10:22:06.145 sclang[70821:86146849] ApplePersistence=NO
sc3> a = FunctionList.new;
-> a FunctionList
sc3> f = { |b| b.postln };
-> a Function
sc3> a.removeFunc(f)
ERROR: Message 'at' not understood.
RECEIVER:
   nil
ARGS:
   Integer 0
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand
		arg this = nil
		arg selector = 'at'
		arg args = [*1]
	Interpreter:interpretPrintCmdLine
		arg this = <instance of Interpreter>
		var res = nil
		var func = <instance of Function>
		var code = "a.removeFunc(f)"
		var doc = nil
		var ideClass = nil
	Process:interpretPrintCmdLine
		arg this = <instance of Main>
^^ ERROR: Message 'at' not understood.
RECEIVER: nil


sc3> a.addFunc(f)
-> a FunctionList
sc3> a.removeFunc(f)
-> nil
sc3> a.removeFunc(f)
-> nil
```

## Expected vs. actual behavior

Calling `removeFunc` before `addFunc` when using FunctionList directly causes an exception. Discussed at [scsynth forum](https://scsynth.org/t/functionlist-fails-if-removing-before-adding/6847) and it seems to be a bug.

fix could be change the `FunctionList` constructor

from:

```
FunctionList : AbstractFunction {
	var <>array, <flopped=false;

	*new { arg functions;
		^super.newCopyArgs(functions)
	}
```

to

```
FunctionList : AbstractFunction {
	var <>array, <flopped=false;

	*new { arg functions;
		^super.newCopyArgs(functions ?? {[]})
	}
```

Workaround I'm using is `a = FunctionList.new([])` but the example in the [documentation](http://doc.sccode.org/Classes/FunctionList.html) uses FunctionList directly without empty array hence other folks might encounter the same issue.

**Comment by jamshark70** [comment]
I believe the idea behind the current implementation is that an empty FunctionList doesn't really need the array until something is added to it.

That is, currently, creating an empty FunctionList creates just one object, but if `array` is eagerly initialized, then the empty FunctionList really creates two objects: a FunctionList and anArray -- i.e. (slightly) heavier garbage-collection load.

This is probably negligible, but, if I remember correctly, I think this was a deliberate decision and we probably shouldn't reverse it casually. (But, if others feel that it's inelegant to initialize the array lazily, I won't object.)

An alternate fix would be:

```
	removeFunc { arg function;
		array.remove(function);
		if(array.size < 2) { ^array.tryPerform(\at, 0) };
	}
```

... where `nil.tryPerform(\at, 0)` is just nil.

**Comment by telephon** [comment]
I think we can afford to just give it an empty array. Then the code is easier to reason about.

**Comment by jamshark70** [comment]
> I think we can afford to just give it an empty array.

Sure, OK. I think very few users are creating a FunctionList and keeping it empty for long periods of time, so the array will be initialized more or less immediately in most cases anyway.

**Comment by gorenje** [comment]
I have no idea about the GC and SC interpreter, I'm only guessing here, however from a performance perspective, won't the `^super.newCopyArgs(functions ?? {[]})` (as suggested by [jordan](https://scsynth.org/t/functionlist-fails-if-removing-before-adding/6847/8)) approach be the best? An empty array would create GC load but a `tryPerform` each time `removeFunc` is called would probably have a greater impact on the SC interpreter.

**Comment by telephon** [comment]
> I have no idea about the GC and SC interpreter, I'm only guessing here, however from a performance perspective, won't the `^super.newCopyArgs(functions ?? {[]})`

yes, that was what I meant …

**Comment by jamshark70** [comment]
At risk of being pedantic, perhaps `functions ?? { Array.new }`? I've grown allergic over the years to strings of punctuation. They're cute but less readable.

**Comment by telephon** [comment]
yes, better.

**Comment by gorenje** [comment]
I was thinking of creating a PR for this and took the easy route of simply creating a `.sclang_conf.yaml` as [described here](https://github.com/supercollider/supercollider/wiki/Setting-up-the-IDE-for-easy-contribution). I forked and cloned the repo, setup my sclang_conf and fired up the interpreter but failed with the unit tests:

sclang_conf:
```
includePaths:
  - /Users/user/supercollider/SCClassLibrary
  - /Users/user/supercollider/testsuite
excludePaths:
  - /Applications/SuperCollider-3.12.2.app/Contents/Resources/SCClassLibrary
  - "/Users/user/Library/Application Support/SuperCollider/Extensions"
```

I checked out the `develop` branch but that won't work (`ERROR: Variable 'rescheduledTime' not defined.`), so then I took the 3.12 branch and ran `/Applications/SuperCollider-3.12.2.app/Contents/MacOS/sclang -l .sclang_conf.yaml` and that worked.

Then I read about the [unit tests](https://github.com/supercollider/supercollider/wiki/Unit-Testing-Guide) and wanted to run them all: `UnitTest.runAll` but that ended in:

```
RUNNING UNIT TEST 'TestDocument'

FAIL: a TestDocument: test_document_getText_retrievesText - Document tests are valid only when run in the IDE

Execution warning: Class 'Document' not found
WARNING: keyword arg 'string' not found in call to Object:doesNotUnderstand
```

I checked my installed 3.12.2 version and it has the same error `sc3> Document.new / ERROR: Class not defined.`

What's the correct procedure here?

**Comment by dyfer** [comment]
@gorenje you can run tests from the IDE, doesn't need to be directly in the interpreter. In the CI we indeed run tests without the IDE, but we run them with [qpm](https://github.com/supercollider/qpm). It allows to control which tests are run (i.e. we don't run the `Document` tests since we don't run the IDE), but I would not recommend setting that up, as it's cumbersome.

Your sclang-conf file looks fine, just make sure in the IDE preferences that the correct file is being used (you can e.g. have multiple files, one with regular config, and one for running tests). 

It's true that with the changes we've made it's impossible to run SC 3.12 with the class library from `develop`. There are two solutions to this:
- either run the develop build of SC which can be downloaded from GitHub Actions or from SC downloads page, or
- switch to the `3.12` branch in the repo and make a new branch based on `3.12` to implement your changes

The latter approach _should_ work, but it might result in issues when merging. If it doesn't work, then you'd need to rebase your PR branch on `develop`, which is an additional step (and after doing that you'd loose the ability to run SC 3.12 with the class library from your PR branch). I'd suggest going with the `develop` build; be aware that you'll need to adjust the `excludePaths` for the appropriate SC version.

**Comment by gorenje** [comment]
It seems that the actions don't generate releases any longer, e.g. https://github.com/supercollider/supercollider/actions/runs/3659996317

I tried creating a dockerfile to create a linux environment to build the development branch. It worked until the test step using [`xvfb-run`](https://github.com/supercollider/supercollider/blob/develop/.github/workflows/actions.yml#L866). It was based on ubuntu 20.04. I then [noticed](https://github.com/supercollider/supercollider/blob/develop/.github/workflows/actions.yml#L794-L797) that for the Linux tests, ubuntu 18.04 and gcc-10 is used. I then tried building a docker image using ubuntu 18.04 and discovered they no longer support gcc-10 any longer, gotta be 20.04! 

Will keep trying to get something working in a docker image!

**Comment by gorenje** [comment]
I've updated the [gist](https://gist.github.com/gorenje/10ea2903a55f395f3e436f08abe9e402) and it now works!

```
3826 TESTS PASSED (total duration: 514.0s)
```

**Comment by dyfer** [comment]
> It seems that the actions don't generate releases any longer, e.g. https://github.com/supercollider/supercollider/actions/runs/3659996317

Builds when pushing to `develop` don't generate releases, but artifacts are available there, on the actions page.


> I tried creating a dockerfile to create a linux environment to build the development branch. It worked until the test step using [`xvfb-run`](https://github.com/supercollider/supercollider/blob/develop/.github/workflows/actions.yml#L866). It was based on ubuntu 20.04. I then [noticed](https://github.com/supercollider/supercollider/blob/develop/.github/workflows/actions.yml#L794-L797) that for the Linux tests, ubuntu 18.04 and gcc-10 is used. I then tried building a docker image using ubuntu 18.04 and discovered they no longer support gcc-10 any longer, gotta be 20.04!

I encountered issues when trying to move tests to Ubuntu 20.04, that's why it's still on 18.04. Eventually I'd like to allow running configurable testsuite without qpm, which would make task of setting up testsuite easier.

**Comment by gorenje** [comment]
> Builds when pushing to develop don't generate releases, but artifacts are available there, on the actions page.

Do you mean the [cache page](https://github.com/supercollider/supercollider/actions/caches)? I can't seem to download the artifacts from there (sorry for my stupidity on this 😄 )

> I encountered issues when trying to move tests to Ubuntu 20.04, that's why it's still on 18.04.

The docker image uses 20.04 + clang - so it does seem doable ... ;)

---

Unfortunately the CI failed on the PR and for me, at two random spots:

- [TestUGen_RTAlloc:test_FFT](https://github.com/supercollider/supercollider/actions/runs/3795175849/jobs/6457342840#step:9:20525) which for me looks like a [timing issue](https://github.com/supercollider/supercollider/blob/c1fb0c5251c0c1fbbc3e951f92a894eff0d82c07/testsuite/classlibrary/TestUGen_RTAlloc.sc#L273-L276) in that the runtime was unexpectedly longer
- [TestServer_boot:test_waitForBoot](https://github.com/supercollider/supercollider/actions/runs/3795175849/jobs/6457342840#step:9:21192) which seems to assume [1 second wait](https://github.com/supercollider/supercollider/blob/1bbe2849a25b83eca2aea8f4e2be0c53a3957df1/testsuite/classlibrary/TestServer_boot.sc#L43-L45) is long enough

It was only the MacOS tests that failed, Linux tests all went through.

Or is there something else going on? I am only making guesses here :-/ Can I do something to prevent/fix this?

---

Btw another question, is there a reason why the asserts in [TestUGen_RTAlloc.sc](https://github.com/supercollider/supercollider/blob/c1fb0c5251c0c1fbbc3e951f92a894eff0d82c07/testsuite/classlibrary/TestUGen_RTAlloc.sc) are done like this:

```
if (completed.isNil) {
    this.assert(false, "failed alloc test should complete");
};
```

and not (i.e. using the inverted if-condition as the assert argument):

```
this.assert(completed.notNil, "failed alloc test should complete");
```

for example?

**Comment by dyfer** [comment]
> > Builds when pushing to develop don't generate releases, but artifacts are available there, on the actions page.
> 
> Do you mean the [cache page](https://github.com/supercollider/supercollider/actions/caches)? I can't seem to download the artifacts from there (sorry for my stupidity on this 😄 )

No worries. They are on the Actions page, but a little bit hidden - one needs to scroll the page down (past the scroll position of embedded actions tree...):
<img width="1284" alt="Screen Shot 2022-12-29 at 17 28 35" src="https://user-images.githubusercontent.com/1589177/209982036-a068518b-6aaa-4d47-bfa1-0aed59703255.png">

(make sure you are in the "Summary" view)


> The docker image uses 20.04 + clang - so it does seem doable ... ;)

Thanks, good to know.

> Unfortunately the CI failed on the PR and for me, at two random spots:

TL;DR you probably don't need to worry about tests completely unrelated to your changes that tail (though sometimes there might be a hidden dependency of a given test on a change elsewhere in the class library, so it's good to be vigilant).

Unfortunately our TestSuite is not "bulletproof" (yet...). Some tests fail randomly (i.e. they fail relatively infrequently, but not due to an actual error). I believe that `TestServer_boot:test_waitForBoot` is one of those. An issue with RtAlloc tests is reported in #5791 I believe. There's also #5359, #5360, #5421.

> Btw another question, is there a reason why the asserts in [TestUGen_RTAlloc.sc](https://github.com/supercollider/supercollider/blob/c1fb0c5251c0c1fbbc3e951f92a894eff0d82c07/testsuite/classlibrary/TestUGen_RTAlloc.sc) are done like this:
(...) 
> for example?

I believe the idea is to not call `.assert` at all if `completed` is not nil at that point.
